lib/ansible/executor/powershell/psrp_put_file.ps1

```diff
@@ -0,0 +1,122 @@
+
+using namespace System.IO
+using namespace System.Reflection
+using namespace System.Security.Cryptography
+
+[CmdletBinding()]
+param (
+    [Parameter(Mandatory = $true)]
+    [string]
+    $Path,
+
+    [Parameter(Mandatory, ValueFromPipeline)]
+    [AllowEmptyString()]
+    [string]
+    $InputObject
+)
+
+begin {
+    $fd = [File]::Create($Path)
+    $algo = [SHA1]::Create()
+    $bytes = @()
+
+    $bindingFlags = [BindingFlags]'NonPublic, Instance'
+    Function Get-Property {
+        <#
+        .SYNOPSIS
+        Gets the private/internal property specified of the object passed in.
+        #>
+        Param (
+            [Parameter(Mandatory = $true, ValueFromPipeline = $true)]
+            [Object]
+            $Object,
+
+            [Parameter(Mandatory = $true, Position = 1)]
+            [String]
+            $Name
+        )
+
+        process {
+            $Object.GetType().GetProperty($Name, $bindingFlags).GetValue($Object, $null)
+        }
+    }
+
+    Function Set-Property {
+        <#
+        .SYNOPSIS
+        Sets the private/internal property specified on the object passed in.
+        #>
+        Param (
+            [Parameter(Mandatory = $true, ValueFromPipeline = $true)]
+            [Object]
+            $Object,
+
+            [Parameter(Mandatory = $true, Position = 1)]
+            [String]
+            $Name,
+
+            [Parameter(Mandatory = $true, Position = 2)]
+            [AllowNull()]
+            [Object]
+            $Value
+        )
+
+        process {
+            $Object.GetType().GetProperty($Name, $bindingFlags).SetValue($Object, $Value, $null)
+        }
+    }
+
+    Function Get-Field {
+        <#
+        .SYNOPSIS
+        Gets the private/internal field specified of the object passed in.
+        #>
+        Param (
+            [Parameter(Mandatory = $true, ValueFromPipeline = $true)]
+            [Object]
+            $Object,
+
+            [Parameter(Mandatory = $true, Position = 1)]
+            [String]
+            $Name
+        )
+
+        process {
+            $Object.GetType().GetField($Name, $bindingFlags).GetValue($Object)
+        }
+    }
+
+    # MaximumAllowedMemory is required to be set to so we can send input data that exceeds the limit on a PS
+    # Runspace. We use reflection to access/set this property as it is not accessible publicly. This is not ideal
+    # but works on all PowerShell versions I've tested with. We originally used WinRS to send the raw bytes to the
+    # host but this falls flat if someone is using a custom PS configuration name so this is a workaround. This
+    # isn't required for smaller files so if it fails we ignore the error and hope it wasn't needed.
+    # https://github.com/PowerShell/PowerShell/blob/c8e72d1e664b1ee04a14f226adf655cced24e5f0/src/System.Management.Automation/engine/serialization.cs#L325
+    try {
+        $Host | Get-Property 'ExternalHost' |
```

Comment:
I know this one's been a thorn for a long time- seems like this would be easier to do ala the *really* old way with multiple round-trips, like WinRM did before I started- should be way faster than WinRM was since it's all happening in the same process/runspace. Not something we need to solve now, but I've got a few ideas that shouldn't require the cleverness or having the whole thing in memory at once.

Reply:
The initial `psrp` code was doing something similar to that but I ended up going the reflection route because it was actually quite slow. The issue is that we needed to setup a brand new pipeline once we reached the buffer limit and that made it slower than the raw input method we use in `winrm` and now with `psrp`.

The code doesn't store it in memory all at once but rather sets the remote runspace to not have a max input data size which is accumulated with each pipeline input. It still "streams" the data but now without it failing once it reaches the max value set internally.